### DISCUSSION TITLE
Expert rules constraint checking

### DISCUSSION THREAD

**Discussion by BegBlev** [netbox-community/netbox#6256] [Ideas]
Hello,
My team is managing a "little" data centre (600 servers and 100 network equipments).
We have a set of business rules:
- if the system is a Dell R640 then its name should be something like "r640-{num}"
- if the system is active it must have an IP assigned.
- if the system is active and it has an IP then it should answer to pings
- the system should have a serial except if it is "planned"
- ...

We have tens of such rules.

Currently we have a python script that checks those rules. But this is only Python test code (if, then, else...) and we are facing 2 problems:
- the code becomes difficult to read/write with lots of nested rules.
- each time we add a rule, we have to modify the code.

I am thinking to use an "expert system". I saw some solutions "Pyke, experta..." but I don't really see how to start.

I know this is not Netbox directly related topic, but I am sure lot of people using Netbox have the same concern and I think that, if the rules could be split apart from the code, they could be integrated inside Netbox in the future.

My questions are:
- do, some of you, make such rules verifications?
- how do you do it?

Best regards
Vincent

**Comment by candlerb** [comment]
> Currently we have a python script that checks those rules.

Is this in the form of a netbox Report, or somewhere else (and if so, where?)

**Reply by BegBlev** [reply]
No, sorry, it was not clear... in our team we have a script for doing this check.

**Reply by candlerb** [reply]
And the script is run periodically to check for anomalies after the fact? Or is it hooked into the GUI or API flow somehow?

**Reply by BegBlev** [reply]
The script is run periodically. Currently this is done "outside" of Netbox... only Netbox API calls. We grab information from other sources too (Vault, iDRAC, iLO, Junos...) and we try to check coherency (i.e : Netbox device serial and Redfish device serial).
This is the same idea as LLDP Netbox view where Netbox cabling is compared to the actual device LLDP database.
I would like a more general approach to check coherency:

- inside 1 data source (i.e if device's Netbox status field is "active" then Netbox "serial" field should be defined.
- between data sources (i.e: device's Netbox serial field should be the same as Redfish serial field)

**Reply by candlerb** [reply]
Thanks for the background.

Taking a Netbox-centric view, I think this could be done using Netbox reports - they can access the ORM data model directly, which is more flexible than the REST API (e.g. you can do joins, yay!)  You can organize reports as one report per business rule.  Being Python they can also call out to external data sources or APIs: some people write reports which ping remote targets or scrape ARP tables, for example.  You can run them [programatically](https://netbox.readthedocs.io/en/stable/additional-features/reports/#via-the-cli), and you get a nice UI for the results.  Some samples [here](https://github.com/netbox-community/reports/tree/master/reports).

Otherwise, your less closely-coupled architecture sounds perfectly reasonable, where Netbox is just one of many systems which a central system cross-references - but I don't have any rules-based system I can recommend.

I am aware that there are products that describe themselves as a "configuration management database" (CMDB), which sync into other systems and provide a central view of that data, and presumably have ways to highlight and resolve inconsistencies between those systems.  I'm not sure it makes sense to buy or build a separate system like that, if for your use case Netbox were able to fulfil that role by itself.

**Reply by BegBlev** [reply]
Thanks for this reply it gives me lots of ideas :)
I will have a look to those "CMDB".
Regarding results presentation, yes, reports seem to be a good option. I will have a look to.
Thank you
